### PR TITLE
ONME-3859 Update network interface to have new address and interface …

### PR DESCRIPTION
**Pull request by tymoteuszblochmobica**
### Description

 ONME-3859 Update network interface to have new address and interface name interfaces
 ONME-3858 Implement socket interface binding to device (SO_BINDTODEVICE) to lwip
To do : ip_route and DNS

### Pull request type

 
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### CODE REVIEW

**Comment by tymoteuszblochmobica** [comment]
@seppotakalo please review
@mikaleppanen please review

**Comment by tymoteuszblochmobica** [comment]
@kjbracey-arm please review

**Comment by SeppoTakalo** [comment]
Cellular team:
@AriParkkila
@mirelachirica

Please review.

**Comment by mikaleppanen** [comment]
As a general comment here I would prefer that the bind to device socket option would be intercepted by socket base class before forwarding it to stack and the interface name would be stored to socket.

Then if the interface name is set, the automatic get host by name operations in tcp connect, udp socket etc. would use the name if it is set.

If tcp connect etc. functions are modified for this feature, the original functions must remain as they are and new functions with interface name need to be created. This includes all the socket classes. 

@SeppoTakalo @kjbracey-arm what do you think?

File: features/netsocket/NetworkInterface.h [hunk 1]
```diff
@@ -160,7 +160,25 @@ class NetworkInterface: public DNS {
      *  @return         NSAPI_ERROR_OK on success, negative error code on failure.
      */
     virtual nsapi_error_t gethostbyname(const char *host,
-                                        SocketAddress *address, const char *interface_name, nsapi_version_t version = NSAPI_UNSPEC);
+                                        SocketAddress *address, nsapi_version_t version = NSAPI_UNSPEC);
+
+    /** Translate a hostname to an IP address with specific version using network interface name.
+         *
```
Comment (mikaleppanen) [inline features/netsocket/NetworkInterface.h]: Check the indentations.

File: features/netsocket/UDPSocket.cpp [hunk 2]
```diff
@@ -37,10 +37,17 @@ nsapi_error_t UDPSocket::connect(const SocketAddress &address)
...
 
-nsapi_size_or_error_t UDPSocket::sendto(const char *host, uint16_t port, const void *data, nsapi_size_t size, const char *interface_name)
+nsapi_size_or_error_t UDPSocket::sendto(const char *host, uint16_t port, const void *data, nsapi_size_t size)
 {
     SocketAddress address;
-    nsapi_size_or_error_t err = _stack->gethostbyname(host, &address, interface_name);
+    nsapi_size_or_error_t err;
+
+    if (strcmp(_interface_name, "")) {
+       	err = _stack->gethostbyname(host, &address);
```
Comment (mikaleppanen) [inline features/netsocket/UDPSocket.cpp]: Check indentation.

File: features/netsocket/DNS.h [hunk 3]
```diff
@@ -35,8 +35,25 @@ class DNS {
...
+     *  The hostname may be either a domain name or an IP address. If the
+     *  hostname is an IP address, no network transactions will be performed.
+     *
+     *  If no stack-specific DNS resolution is provided, the hostname
+     *  will be resolve using a UDP socket on the stack.
+     *
+     *  @param host     Hostname to resolve.
+     *  @param address  Pointer to a SocketAddress to store the result.
+     *  @param version  IP version of address to resolve, NSAPI_UNSPEC indicates
+     *  @param interface_name  Network interface_name
```
Comment (mikaleppanen) [inline features/netsocket/DNS.h]: There is also this version of the parameter in commit:
@param interface_name Network interface name
Change all to same format... maybe the "Network interface name" is better.

File: features/netsocket/NetworkStack.h [hunk 4]
```diff
@@ -43,7 +43,7 @@ class NetworkStack: public DNS {
      *  @return         Null-terminated representation of the local IP address
      *                  or null if not yet connected
      */
```
Comment (mikaleppanen) [inline features/netsocket/NetworkStack.h:46]: Description for new parameter missing.

File: features/netsocket/NetworkStack.h [hunk 5]
```diff
@@ -116,7 +161,7 @@ class NetworkStack: public DNS {
      *  @param address  Destination for the host address
      *  @return         NSAPI_ERROR_OK on success, negative error code on failure
      */
-    virtual nsapi_error_t add_dns_server(const SocketAddress &address);
+    virtual nsapi_error_t add_dns_server(const SocketAddress &address, const char *interface_name);
```
Comment (mikaleppanen) [inline features/netsocket/NetworkStack.h]: Should default to to = NULL, parameter description missing

File: features/netsocket/NetworkStack.h [hunk 6]
```diff
@@ -127,7 +172,7 @@ class NetworkStack: public DNS {
      *  @param address  Destination for the host address
      *  @return         NSAPI_ERROR_OK on success, negative error code on failure
      */
-    virtual nsapi_error_t get_dns_server(int index, SocketAddress *address);
+    virtual nsapi_error_t get_dns_server(int index, SocketAddress *address, const char *interface_name);
```
Comment (mikaleppanen) [inline features/netsocket/NetworkStack.h]: Should default to to = NULL, parameter description missing

File: features/netsocket/OnboardNetworkStack.h [hunk 7]
```diff
@@ -103,7 +108,7 @@ class OnboardNetworkStack : public NetworkStack {
          * @param    buflen     size of supplied buffer
          * @return              Pointer to a buffer, or NULL if the buffer is too small
          */
-        virtual char *get_ip_address(char *buf, nsapi_size_t buflen) = 0;
+        virtual char *get_ip_address(char *buf, nsapi_size_t buflen, const char *interface_name) = 0;
```
Comment (mikaleppanen) [inline features/netsocket/OnboardNetworkStack.h]: Should default to to = NULL, parameter description missing

File: features/netsocket/InternetSocket.cpp [hunk 8]
```diff
@@ -171,6 +171,9 @@ nsapi_error_t InternetSocket::setsockopt(int level, int optname, const void *opt
         ret = NSAPI_ERROR_NO_SOCKET;
     } else {
         ret = _stack->setsockopt(_socket, level, optname, optval, optlen);
+        if(optname == NSAPI_BIND_TO_DEVICE) {
```
Comment (mikaleppanen) [inline features/netsocket/InternetSocket.cpp]: Also level should be checked.

File: features/netsocket/InternetSocket.cpp [hunk 9]
```diff
@@ -171,6 +171,9 @@ nsapi_error_t InternetSocket::setsockopt(int level, int optname, const void *opt
...
+        	strncpy(_interface_name,static_cast<const char*>(optval),3);
```
Comment (mikaleppanen) [inline features/netsocket/InternetSocket.cpp]: Do not copy more than optlen

File: features/lwipstack/LWIPStack.cpp [hunk 10]
```diff
@@ -508,6 +509,24 @@ nsapi_error_t LWIP::setsockopt(nsapi_socket_t handle, int level, int optname, co
     struct mbed_lwip_socket *s = (struct mbed_lwip_socket *)handle;
 
     switch (optname) {
+    	case NSAPI_BIND_TO_DEVICE:
+             if (optlen != sizeof(char[2])) {
```
Comment (mikaleppanen) [inline features/lwipstack/LWIPStack.cpp]: I think we do not want to fix the length. There should be maximum length and name is ok if it is shorter.

File: features/netsocket/nsapi_types.h [hunk 11]
```diff
@@ -258,6 +263,7 @@ typedef enum nsapi_socket_option {
     NSAPI_RCVBUF,            /*!< Sets recv buffer size */
     NSAPI_ADD_MEMBERSHIP,    /*!< Add membership to multicast address */
     NSAPI_DROP_MEMBERSHIP,   /*!< Drop membership to multicast address */
+	NSAPI_BIND_TO_DEVICE, 		 /*!< Bind socket network interface name*/
```
Comment (mikaleppanen) [inline features/netsocket/nsapi_types.h]: indentation

File: features/netsocket/nsapi_dns.h [hunk 12]
```diff
@@ -91,6 +105,21 @@ nsapi_error_t nsapi_dns_query_async(NetworkStack *stack, const char *host,
...
+ *  @param callback Callback that is called for result
+ *  @param interface_name Network interface name
+ *  @param version  IP version to resolve (defaults to NSAPI_IPv4)
+ *  @return         0 on success, negative error code on failure or an unique id that
+ *                  represents the hostname translation operation and can be passed to
+ *                  cancel, NSAPI_ERROR_DNS_FAILURE indicates the host could not be found
+ */
+nsapi_error_t nsapi_dns_query_async(NetworkStack *stack, const char *host,
+                                    NetworkStack::hostbyname_cb_t callback, call_in_callback_cb_t call_in_cb,
+									const char *interface_name, nsapi_version_t version = NSAPI_IPv4);
```
Comment (mikaleppanen) [inline features/netsocket/nsapi_dns.h]: indentation

**Review by mikaleppanen — CHANGES_REQUESTED** [review]

File: features/netsocket/NetworkInterface.h (hunk 1 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/NetworkInterface.h]: Done

File: features/netsocket/UDPSocket.cpp (hunk 2 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/UDPSocket.cpp]: Done

File: features/netsocket/DNS.h (hunk 3 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/DNS.h]: Done

File: features/netsocket/NetworkStack.h (hunk 4 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/NetworkStack.h:46]: Done

File: features/netsocket/NetworkStack.h (hunk 5 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/NetworkStack.h]: Done

File: features/netsocket/NetworkStack.h (hunk 6 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/NetworkStack.h]: Done

File: features/netsocket/OnboardNetworkStack.h (hunk 7 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/OnboardNetworkStack.h]: Done

File: features/netsocket/InternetSocket.cpp (hunk 8 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/InternetSocket.cpp]: Done

File: features/netsocket/InternetSocket.cpp (hunk 9 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/InternetSocket.cpp]: Done

File: features/lwipstack/LWIPStack.cpp (hunk 10 above)
Comment (tymoteuszblochmobica) [reply features/lwipstack/LWIPStack.cpp]: Done

File: features/netsocket/nsapi_types.h (hunk 11 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/nsapi_types.h]: Done

File: features/netsocket/nsapi_dns.h (hunk 12 above)
Comment (tymoteuszblochmobica) [reply features/netsocket/nsapi_dns.h]: Done

**Pull request closed without merging**